[REVISIT] Allow deprecating an entire package by deprecating its package object [ci: last-only]

som-snytt opened the pull request:
Package members are deprecated if the package object of an enclosing package is deprecated.

Fixes https://github.com/scala/bug/issues/3115

SethTisue commented:
@som-snytt are you wanting on feedback on this yet?

som-snytt commented:
@SethTisue I don't remember, but if it works, it would be cool. I'll rebase it and see.

som-snytt commented:
I guess this UI reports a conflict because this branch had an old commit that has been merged since then.

som-snytt commented:
> Semantics: suppress deprecation warnings; induce deprecation warnings.

Poetic! I wonder what I meant.

lrytz commented on src/compiler/scala/tools/nsc/typechecker/Typers.scala:
```diff
@@ -3918,8 +3918,14 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
-          runReporting.addSuppression(Suppression(info.pos, List(MessageFilter.Category(WarningCategory.Deprecation)), start, end, synthetic = true))
+          def suppressAt(start: Int, end: Int): Unit =
+            runReporting.addSuppression(Suppression(info.pos, List(MessageFilter.Category(WarningCategory.Deprecation)), start, end, synthetic = true))
+          rangeFinder() match { case (start, end) => suppressAt(start, end) case _ => }
```
`rangeFinder` always returns a pair, why this match?

som-snytt replied:
weekend coding.

lrytz commented on src/compiler/scala/tools/nsc/typechecker/Typers.scala:
```diff
@@ -3918,8 +3918,14 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
+          val annsym = annotee.get.symbol
+          if (annsym != null && annsym.isPackageObject && annsym.owner != null && annsym.owner.pos.isDefined) {
+            val p = annsym.owner.pos
+            suppressAt(p.start, p.end)
```
this only works if rangepos is enabled

som-snytt replied:
please don't crush my dreams.

som-snytt replied:
I'll try to come up with a different idea.

lrytz commented on src/compiler/scala/tools/nsc/typechecker/RefChecks.scala:
```diff
@@ -1261,19 +1261,14 @@ abstract class RefChecks extends Transform {
+      // If symbol is deprecated and the point of reference is not enclosed in a deprecated member, issue a warning.
+      // Or, if the symbol is not directly deprecated but is a top-level member of some deprecated packaging, then warn.
+      if (sym.isDeprecated && !currentOwner.ownersIterator.exists(x => x.isDeprecated || x.packageObject.isDeprecated) ||
+          !sym.isDeprecated && sym.isTopLevel && sym.ownersIterator.exists(x => x.isDeprecated || x.packageObject.isDeprecated))
```
does `x.isDeprecated` ever fire here?

som-snytt replied:
You're like my favorite TA.

lrytz commented on test/files/neg/t3115b.check:
```diff
@@ -0,0 +1,9 @@
+t3115b.scala:17: warning: class C in package p is deprecated
```
that's not entirely accurate...

som-snytt replied:
I didn't think yet to update the messaging.

som-snytt replied:
Edit: "Oh so now the error messages have to be accurate."

som-snytt committed:
Deprecation of package members

Annotation applied to nested object package
deprecates all top-level definitions in nested packages.
That is, if a reference requires a path using
the deprecated package, then the element is deprecated.
References in the scope of the deprecation receive
the standard reprieve.

som-snytt committed:
Avoid NoSymbol check under -Xdev

som-snytt commented:
Here is a straightforward approach since Lukas sent me back to the drawing board last spring. It seems to work, so I must be doing something wrong. The limitations are that top-level packages cannot be annotated, and a deprecation of package object cannot suppress syntax deprecations.

That solution should perhaps be reworked: instead of suppressing by range, use positions later to locate the elements involve. For example, after typer, re-process deprecated `then` by locating the tree and its enclosing element so refchecks can properly consult symbol info, instead of allows issuing and then suppressing by position.

SethTisue commented:
@lrytz want to take a second look at this?

som-snytt commented:
@SethTisue Some of us have lives! Some of us have moved on to Scala 3!

som-snytt commented:
Thanks, was about to look at your comment yesterday. Today is U.S. holiday for which I am on the hook to do all the cooking, so it may not happen today. I'll re-draft to make it obvious to future selves.

som-snytt committed:
Use package deprecation message

som-snytt commented:
It's been almost nine months, the proper gestational period. Ready to deliver!

Added a clause that it is "deprecated in package p", plus the expected message.

I noticed that `sym.locationString` is commented "Doesn't do much, for backward compatibility reasons." Not sure what that means, but it would be nice to have full paths, "deprecated in package p.q".

som-snytt committed:
Inline ownsString

SethTisue commented:
@lrytz re-review?

SethTisue commented:
I hadn't understood the purpose of the PR but Lukas explained — I've changed the title.

I worry a bit that there might be a library author out there who has a deprecated package object that they didn't intend in this way. Maybe I worry too much...?

SethTisue commented:
If Scala 2 is to behave this way, Scala 3 should as well. I wonder if there's any possibility of pushback or whether this sufficiently uncontroversial.

SethTisue commented:
@som-snytt did you consider changing the parser so that `@deprecated package object` would be allowed and users wouldn't have to resort to the `package p { @deprecated object `package` ...` thing?

som-snytt commented:
I don't remember how this PR played out, but I'll find time to refresh my memory, or what's left of it.

Scala 3 eliminates package objects, so it would require a different idiom.

Since Dotty frees up some uses of underscore, obviously:

```
package p:
  @deprecated("everything", since="yesterday") _
```

SethTisue commented:
> Scala 3 eliminates package objects, so it would require a different idiom.

Ah, good point. Not sure this is worth doing unless we can find a cross-buildable syntax. (And if we cannot find such a syntax, then maybe it should just be a Scala 3 feature.)
